pstoreds/addr_book.go

```diff
@@ -2,398 +2,306 @@ package pstoreds
-	mh "github.com/multiformats/go-multihash"
-
 	peer "github.com/libp2p/go-libp2p-peer"
 	pstore "github.com/libp2p/go-libp2p-peerstore"
+	pb "github.com/libp2p/go-libp2p-peerstore/pb"
 	pstoremem "github.com/libp2p/go-libp2p-peerstore/pstoremem"
-)
 
-var (
-	log = logging.Logger("peerstore/ds")
-	// The maximum representable value in time.Time is time.Unix(1<<63-62135596801, 999999999).
-	// But it's too brittle and implementation-dependent, so we prefer to use 1<<62, which is in the
-	// year 146138514283. We're safe.
-	maxTime = time.Unix(1<<62, 0)
-
-	ErrTTLDatastore = errors.New("datastore must provide TTL support")
+	lru "github.com/hashicorp/golang-lru"
+	ma "github.com/multiformats/go-multiaddr"
+	b32 "github.com/whyrusleeping/base32"
 )
 
-// Peer addresses are stored under the following db key pattern:
-// /peers/addr/<b32 peer id no padding>/<hash of maddr>
-var abBase = ds.NewKey("/peers/addrs")
-
-var _ pstore.AddrBook = (*dsAddrBook)(nil)
-
-// dsAddrBook is an address book backed by a Datastore with both an
-// in-memory TTL manager and an in-memory address stream manager.
-type dsAddrBook struct {
-	cache        cache
-	ds           ds.TxnDatastore
-	subsManager  *pstoremem.AddrSubManager
-	writeRetries int
-}
-
 type ttlWriteMode int
 
 const (
 	ttlOverride ttlWriteMode = iota
 	ttlExtend
 )
 
-type cacheEntry struct {
-	expiration time.Time
-	addrs      []ma.Multiaddr
-}
-
-type addrRecord struct {
-	ttl  time.Duration
-	addr ma.Multiaddr
-}
+var (
+	log = logging.Logger("peerstore/ds")
 
-func (ar *addrRecord) MarshalBinary() ([]byte, error) {
-	ttlB := make([]byte, 8)
-	binary.LittleEndian.PutUint64(ttlB, uint64(ar.ttl))
-	return append(ttlB, ar.addr.Bytes()...), nil
-}
+	// Peer addresses are stored db key pattern:
+	// /peers/addrs/<b32 peer id no padding>
+	addrBookBase = ds.NewKey("/peers/addrs")
+)
 
-func (ar *addrRecord) UnmarshalBinary(b []byte) error {
-	ar.ttl = time.Duration(binary.LittleEndian.Uint64(b))
-	// this had been serialized by us, no need to check for errors
-	ar.addr, _ = ma.NewMultiaddrBytes(b[8:])
-	return nil
+// addrsRecord decorates the AddrBookRecord with locks and metadata.
+type addrsRecord struct {
+	sync.RWMutex
+	*pb.AddrBookRecord
+	dirty bool
 }
 
-// NewAddrBook initializes a new address book given a
-// Datastore instance, a context for managing the TTL manager,
-// and the interval at which the TTL manager should sweep the Datastore.
-func NewAddrBook(ctx context.Context, store ds.TxnDatastore, opts Options) (*dsAddrBook, error) {
-	if _, ok := store.(ds.TTLDatastore); !ok {
-		return nil, ErrTTLDatastore
+// flush writes the record to the datastore by calling ds.Put, unless the record is
+// marked for deletion, in which case we call ds.Delete. To be called within a lock.
+func (r *addrsRecord) flush(write ds.Write) (err error) {
+	key := addrBookBase.ChildString(b32.RawStdEncoding.EncodeToString([]byte(r.Id.ID)))
+	if len(r.Addrs) == 0 {
+		return write.Delete(key)
```

Comment:
Should we set `r.dirty = false` here too?

Reply:
Done.